### ISSUE TITLE
epic: Custom Σ₀ local coding agent — Ollama + LoRA trained on Claude sessions

### ISSUE THREAD

**Issue by alex-place** [alex-place/lantern-os#690]
## Epic: Custom Σ₀ local coding agent (Ollama, LoRA trained on Claude sessions)

Goal stated by owner: "Ollama hooked up to our own custom Σ₀ coding-agent LoRA, trained on our past Claude sessions + external grounding, looped per research into the knowledge base and a central CSF store."

This issue tracks turning that from aspiration into reality, with honest status.

### Starting reality (verified 2026-06-18)
- Ollama was **not running**; autowork's coder ran on **Claude/Gemini** with Ollama as an *offline, stock-`qwen2.5-coder`* fallback.
- **No coding LoRA existed** (`models/` empty, `data/training/` empty, `OLLAMA_MODEL` unset).
- The only trained LoRA on disk is the **Three Doors image** adapter (`D:\lantern-lora`, unrelated to coding).
- `lantern-csf-dream` in Ollama is a **prompt-wrapper Modelfile** (dream game), not a fine-tune.

### Done in this pass ✅
- [x] Extract real training data from 51 Claude sessions → `data/training/haiku-ft-pairs.jsonl` (365 high-quality pairs). `scripts/extract-session-pairs.py`
- [x] Convert to alpaca format for QLoRA → `models/lantern-sigma0-coder/training-data.jsonl`. New `scripts/convert-pairs-to-alpaca.py`
- [x] **Ollama now live** (`ollama serve`, port 11434) with `qwen2.5-coder` warm.
- [x] Verified autowork's `callOllama` path returns real coding answers against live Ollama.
- [x] Created **`lantern-sigma0-coder`** Ollama model — repo-grounded Σ₀ system prompt over `qwen2.5-coder` (`models/lantern-sigma0-coder/Modelfile`). Verified it answers grounded in the codebase.
- [x] Wired `OLLAMA_MODEL=lantern-sigma0-coder` in `.env.local` so autowork uses it.

> Honesty: `lantern-sigma0-coder` is currently a **prompt-grounded** model, NOT yet a fine-tuned LoRA. It is a real, usable custom coding model — but the "trained on Claude sessions" weights step is below.

### Remaining ⏳ (needs GPU + unsloth)
- [ ] Install training stack: `pip install unsloth` (Windows is finicky; WSL/Linux recommended). GPU present (`nvidia-smi` OK).
- [ ] Run QLoRA training: `python scripts/fine-tune-ollama-model.py --model lantern-sigma0-coder --data models/lantern-sigma0-coder/training-data.jsonl --epochs 3`
- [ ] Export adapter → GGUF, update `models/lantern-sigma0-coder/Modelfile` `FROM` the fine-tuned GGUF (Stage 2 noted in the Modelfile), `ollama create` again.
- [ ] Confirm autowork end-to-end uses the fine-tuned model on a real issue.

### Related / sub-tracks
- External grounding + research loop: partially real today — autowork does `git grep` codebase research + DuckDuckGo web search + writes convergence records (`data/convergence-autonomous-work.jsonl`). Harden into a retrieval loop over the CSF store. (#629)
- Central CSF store: `data/csf/` exists but is minimal; wire it as a retrieval source for grounding. 
- Short `callOllama` timeout (30s) trips on cold model load — bump or pre-warm. (found this pass)
- Relates to #628 (Local Ollama coding agent), #629 (Ollama agent failure diagnosis).

### Privacy
Training JSONL is derived from Claude session transcripts — **must stay gitignored / out of the repo** (split-data-model). Only scripts + Modelfile are committed.

**Comment by alex-place** [comment]
### Training-environment blocker (verified 2026-06-18)
- GPU present: **NVIDIA RTX 3070, 8 GB VRAM** (`nvidia-smi` OK).
- But **torch is CPU-only**: `2.10.0+cpu`, `torch.cuda.is_available() == False`.
- `unsloth` not installed.

So the QLoRA step cannot run as-is. To unblock (recommend an **isolated venv** so the running trading service / pytest deps are not disturbed):
```
python -m venv .venv-train && .venv-train\Scripts\activate
pip install torch --index-url https://download.pytorch.org/whl/cu121
pip install unsloth
python scripts/fine-tune-ollama-model.py --model lantern-sigma0-coder --data models/lantern-sigma0-coder/training-data.jsonl --epochs 3
```
Notes:
- 8 GB VRAM is tight — use a <=7B base in 4-bit (qwen2.5-coder-7B QLoRA fits; larger will OOM).
- unsloth on native Windows is finicky; WSL2 + CUDA is the reliable path.
- Everything upstream is ready: 365 alpaca records at `models/lantern-sigma0-coder/training-data.jsonl`, base `lantern-sigma0-coder` registered, Modelfile Stage-2 `FROM` swap documented.

torch+CUDA was NOT reinstalled automatically — it can break the Python env the trading service and tests use. Flagging for an explicit decision / isolated-venv run.

**Comment by alex-place** [comment]
## ✅ DONE — LoRA trained, deployed, wired, verified (2026-06-18)

The claim is now real (with the one honest scope note below).

**Trained:** QLoRA via peft + bitsandbytes (`scripts/train-qlora-peft.py`) on base `Qwen/Qwen2.5-Coder-3B-Instruct` (3B chosen to fit the 8 GB RTX 3070), on **365 pairs extracted from 51 Claude Code sessions**. 3 epochs / 135 steps, **loss 2.87 → 1.78**.

**Deployed:** adapter (119 MB) → merged fp16 (`scripts/merge-lora.py`) → imported to Ollama as **`lantern-sigma0-coder-v2`**. Verified it returns real code; autowork's `callOllama` path reaches it (`AUTOWORK→OLLAMA: "Fine-tuned weights for specific task."`).

**Wired (persisted):** `OLLAMA_MODELS=D:\ollama-models` (User env), `OLLAMA_MODEL=lantern-sigma0-coder-v2` (`.env.local`). Ollama store relocated to D: because C: was full (0.1 GB free); all prior models copied over so nothing was lost.

**On master (`f0f6933f`):** training/merge scripts, `Modelfile.v2` build record, handoff marked complete. Weights stay off-repo (privacy + size): `D:\lantern-train\` (adapter + merged), `D:\ollama-models\` (GGUF).

### Honest scope notes
- Base is **3B, not 7B** — the 8 GB card can't train 7B QLoRA reliably; 3B is what fits and completes. Bump to 7B on a bigger GPU later if desired.
- This is a **single-pass** train on 365 pairs. Quality is "real LoRA," not "production-grade." Re-train as more sessions accumulate.
- Stock `unsloth` was skipped (broken on native Windows); used the portable peft/bitsandbytes path instead.

### Remaining polish (separate)
- External grounding / research loop into CSF store is still partial (#629).
- `callOllama` 30 s timeout trips on cold model loads — pre-warm or bump.

**Comment by Mookman11** [comment]
## Verified re-check (2026-06-19)

Re-ran the verification on the actual machine. Status has drifted from the "Done ✅" list, and the training blocker has changed — partly for the better. Honest snapshot:

**The Σ₀ work is in a separate checkout.** It lives in `C:\Users\micah\lanternos\` (branch `feat/gameplay-centered`), not the home-root checkout. Sibling checkouts `lantern-os/` (master) and `lantern-workspace/` also exist on disk.

| "Done ✅" claim | Reality on 2026-06-19 |
|---|---|
| Ollama live, `qwen2.5-coder` warm | `ollama serve` **is running** (v0.30.10, port 11434), but `ollama list` is **empty** — no `qwen2.5-coder`, no `lantern-sigma0-coder` |
| `lantern-sigma0-coder` Ollama model created | **Does not exist** in Ollama right now |
| 365 pairs → `data/training/haiku-ft-pairs.jsonl` | **Missing on disk** (gitignored per the privacy rule; never persisted/regenerated) |
| alpaca data → `models/lantern-sigma0-coder/training-data.jsonl` | **Missing on disk** |
| `OLLAMA_MODEL` wired in `.env.local` | **Not set**; no `.env.local` present |
| Modelfile committed | **Exists** in `lanternos/models/lantern-sigma0-coder/Modelfile` (Stage-1 prompt wrapper, `FROM qwen2.5-coder:latest`) |
| extract / convert / fine-tune scripts | All three present in `lanternos/scripts/` |

**Training blocker changed — for the better:**
- Earlier note said **RTX 3070, 8 GB** ("tight, will OOM"). Actual hardware now: **RTX 4080 SUPER, 16 GB** (driver 591.86). That removes the VRAM constraint — qwen2.5-coder-7B QLoRA fits comfortably.
- torch is now **not installed at all** in the default Python (earlier it was `2.10.0+cpu`); `unsloth` not installed. The training env still needs building — in an isolated venv, as flagged.

**Honest bottom line:**
- **Stage-1 is not currently live.** The Ollama model and both data files are gone, so the "real, usable prompt-grounded model" would need rehydrating (regenerate pairs → alpaca, pull `qwen2.5-coder`, `ollama create`, set `OLLAMA_MODEL`). The scripts + Modelfile to do it still exist.
- **Stage-2 (the actual fine-tune) is now unblocked by the GPU upgrade**, but depends on regenerating the training data first, since it was never persisted.

No code or model changes made in this pass — verification only.

<sub>Verified via: `ollama list` / `ollama ps` / `ollama --version`, `nvidia-smi`, filesystem checks of the listed paths across the `lanternos` / `lantern-os` / `lantern-workspace` checkouts, and `python -c "import torch"` (ModuleNotFoundError).</sub>

**Comment by alex-place** [comment]
Closing as not planned / superseded. The custom Σ₀ coding agent is now the Ouro 1.4B LoopLM spiral coder (epic #774; `scripts/ouro_serve.py` + `src/sigma0/loop_lm.py`), not an Ollama+LoRA-on-qwen fine-tune. Ollama has been sunset (#811), so the "trained on Claude sessions" QLoRA-on-qwen sub-track is no longer the plan. Tracking continues under #774.
